Add language: LPC

jlchmura opened the pull request:
## Description
This PR adds syntax highlighting for the LPC language, a C-like language used in MUD drivers such as [LDMud](https://www.ldmud.eu/) and [FluffOS](https://www.fluffos.info/).

## Checklist:
- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:  
      -  https://github.com/search?type=code&q=NOT+is%3Afork+path%3A*.c+%28%22this_player%28%29%22+OR+%22this_object%28%29-%3E%22%29
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - https://github.com/jlchmura/lpc-language-server/blob/main/server/src/tests/cases/compiler/
    - Sample license(s):
      - MIT 
  - [x] I have included a syntax highlighting grammar: https://github.com/jlchmura/lpc-language-server/tree/main/syntaxes
  - [x] I have added a color
    - Hex value: `#ffb100`
    - Rationale: It is reminiscent of the amber screens that were in use when many MUDs got their start
  - [x] I have updated the heuristics to distinguish my language from others using the same extension.

jlchmura commented:
Note to the maintainers: I was not able to write a heuristic that could handle `samples/c/nightmare.h` since C and LPC are so similar in that regard.  Do you recommend removing LPC from the `.h` extension, and let the user override via gitattributes `linguist-language`?

jlchmura commented:
This PR will close #7178

jlchmura commented:
@Alhadis @lildude Can you add this PR to your to-do list to review?

lildude reviewed (changes requested):
Please review your regexes to ensure they run linearly and do not suffer from ReDoS... your `lpc` pattern definitely suffers from ReDoS.

Please also update your sample to only include real-world examples. Several of your samples are contrived just to validate the syntax highlighter. Additionally, and sample that is suppressed in the diff indicates it's too big for our needs so will need removing or reducing.

lildude commented on lib/linguist/heuristics.yml:
```diff
@@ -174,6 +174,12 @@ disambiguations:
+  rules:
+    - language: LPC
+      named_pattern: lpc
+      negative_pattern: '(?i:(?:[ \t]*(?:unsigned|uint|enum)|^\s*\*[A-Za-z]|[A-Za-z][A-Za-z0-9_]*\[\]))'
```
This is vulnerable to ReDos.

lildude reviewed (changes requested):
Your regexes are still vulnerable to ReDoS.

> @lildude I still need advice on `samples/c/nightmare.h`. Remove LPC from the `.h` extension?

Whooops. Sorry I missed that. If you can't write a heuristic that only matches LPC, then it's best to leave it off. We fought for years trying to find the compromise we currently have, and even have it [documented in our troubleshooting guide](https://github.com/github-linguist/linguist/blob/main/docs/troubleshooting.md#my-ccobjective-c-h-header-file-is-detected-as-the-wrong-language). 

The same applies for your `.c` heuristic - it need to 100% only match LPC as I don't want to get into a similar battle with C.

The regexes also need to be safe from ReDos and RE2 compatible.

jlchmura commented:
Thanks @lildude.
I reworked the LPC named pattern to make it even more LPC-specific, and removed the ReDoS-vulnerable negative patterns in the heuristic, which I don't believe are needed anymore.  

Let me know...

lildude reviewed (changes requested):
We need tests for your heuristics and a search query for the `.h` extension in the PR template too.
